feat: add publishing to PyPI step

Alchez opened the pull request:
Closes #21

---

@agritheory, the way to add your PyPI user as a trusted publisher: https://docs.pypi.org/trusted-publishers/creating-a-project-through-oidc/

This would avoid needing us to explicitly add a token inside the Github Action environment.

Here are the variables you'll need to set in https://pypi.org/manage/account/publishing:
- PyPI Project Name: `forecast`
- Owner: `agritheory`
- Repository name: `forecast`
- Workflow name: `release.yaml`
- Environment name: `pypi`

HKuz commented on .github/workflows/tests.yml:
```diff
@@ -19,7 +19,7 @@ jobs:
           restore-keys: ${{ runner.os }}-pip
-  
+
   tests:
```
Since we're supporting 3.10, <3.14, we should add tox tests for 3.13 env. In the `python-version` section, can add `'3.13'`

HKuz commented on pyproject.toml:
```diff
@@ -34,11 +52,14 @@ use_parentheses = true
+requires =
     tox-gh-actions>=3
 envlist = py310, py311, py312, linting
 
```
Under the `python=` section, let's add a line for `3.13: py313, linting` and remove `linting` from `3.12`.

Then change the `[testenv:py312]` key and add a new `[testenv:py313]` one so it looks like:

```
[testenv:py312]
description = run pytest on 3.12
commands =
    pytest tests/ --disable-warnings -s
[testenv:py313]
description = run pytest and coverage on 3.13
commands =
    pytest tests/ --cov=forecast --cov-report=xml --disable-warnings -s
```

And finally - looks like the tests aren't running because the poetry lock file changed. Potential solution may be to add the `poetry check` and `poetry lock` commands in both the `[testenv]` and `[testenv:linting]` `commands_pre` sections?

(GitHub isn't letting me add these comments on the exact line since those lines didn't change, sorry!)

HKuz reviewed (changes requested):
Overall, nice work - I suggested to add python 3.13 to the tox test matrix since we're explicitly supporting >=3.10, <3.14. Because we're adding a more recent Python version, we can use that env for the linting step.

Only other thing is when I search PyPI, I see there's already a package named "forecast" with the `pip install forecast` command. I *believe* that names need to be unique on PyPI, so we'd likely have to use something like "forecast-at". Out of scope for this PR, but whatever the name ends up being, I'll need to update our docs and example notebooks with the new name for the import statements.

Alchez commented:
@agritheory @HKuz I've added the requested changes, and also updated the project to use [Poetry v2](https://python-poetry.org/blog/announcing-poetry-2.0.0/), which brings the `pyproject.toml` definitions for Poetry in harmony with base implementations of that file, and only extends it with the `[tool.poetry]` block.

HKuz reviewed (approved):
Changes I requested look good 👍 (I'm not sure on the answer to changing the repo name itself, but agree that will require code and documentation changes)
